content/blog/r-frictionless-validation.md

```diff
@@ -0,0 +1,85 @@
+```
+
+We can see the row and column with the invalid type, we can also see the actual value and the expected type. This is useful information but it isn't a straightforward message we can surface to the end user. It will still require some work to build an error string from the returned information.
+
+Let's try adding a `constraint` to our table schema to ensure that a column's values match a set of possible values. If we load some invalid data using this schema, again we get a warning which if when we run `vroom::problems()` we can see the details of
+
+```R
+# A tibble: 1 × 5
+     row   col expected           actual   file 
+   <int> <int> <chr>              <chr>    <chr>
+ 1     1     1 value in level set other    ""  
+```
+
+Again `frictionless` tell us the offending row and column and the value found in that column, but it doesn't tell us what possible values it could have taken, making it less actionable. The end user would have to have knowledge of the schema to be able to fix this.
+
+This seems to be about the limit of what the `frictionless` R package can do in terms of validation. They don't support a `validation` function and don't plan on adding it in the future. It looks like the package isn't quite intended for our use case. 
+
+It also does not support some of our other use cases. The `frictionless` R package can only handle tabular data, not geojson. And the current implementation at the time of writing does not support `missingValues` from the [table schema spec](https://specs.frictionlessdata.io/table-schema/#missing-values) which would be essential in R for working with `NA` values.
+
+## datapackage.r
+
+The second package `datapackage.r` is also developed by the Frictionless Data organisation. This is an older package and looks to be out of active development in favour of new `frictionless` pacakge. It does however offer validation by using the `tableschema.r` package.
+
+This works in a similar way to the `frictionless` package to create and modify Frictionless data packages. Like with `frictionless` validation against the schema is done when we read the data. If we try to read a data package with a resource with a missing column we get an error when calling `read()` like
+
+```a
+Error: Table headers don't match schema field names
+```
+
+This also does not look very actionable, which table headers don't match? Are some missing? Are they misspelled? This combined with the fact that the package is not under active development doesn't make this a compelling option.
+
+## Reticulate
+
+Another possibility would be to use reticulate to use the Python validation code directly. The Python implementation does support a wide range of build in checks and has ability to specify [custom checks](https://framework.frictionlessdata.io/docs/guides/validating-data.html#custom-checks). Additionally, it supports validation for JSON files, and other file types can be supported using plugins.
```

Comment:
Also add that one of our requirements is that the R package needs to function in a normal user environment and that adding in reticulate significantly complicates the installation?